clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java

```diff
@@ -103,155 +80,148 @@ public Collection<AgentDataType> getProvidedDataTypes() {
-                  objectMapper.convertValue(metadataAttributes, ATTRIBUTES),
-                  Collections.emptyMap());
-        } else {
-          metadata = null;
-        }
-      }
-      Map<String, Collection<String>> evictions = new HashMap<>();
-      Map<String, Collection<CacheData>> cacheResults = new HashMap<>();
-      List<CacheData> instanceTypeData = new ArrayList<>();
-      cacheResults.put(Keys.Namespace.INSTANCE_TYPES.getNs(), instanceTypeData);
-      if (metadata != null) {
-        cacheResults.put(getAgentType(), Collections.singleton(metadata));
-      } else {
-        evictions.put(getAgentType(), Collections.singleton("metadata"));
-      }
+    if (metadata != null) {
+      metadataAttributes =
+          objectMapper.convertValue(metadata.getAttributes(), MetadataAttributes.class);
+    } else {
+      MetadataAttributes newMetadataAttributes = new MetadataAttributes();
+      newMetadataAttributes.cachedInstanceTypes = instanceTypes;
+      metadataAttributes = newMetadataAttributes;
+    }
 
-      for (String instanceType : instanceTypes) {
-        for (String account : matchingAccounts) {
-          Map<String, Object> instanceTypeAttributes = new HashMap<>();
-          instanceTypeAttributes.put("account", account);
-          instanceTypeAttributes.put("region", region);
-          instanceTypeAttributes.put("name", instanceType);
-          instanceTypeData.add(
-              new DefaultCacheData(
-                  Keys.getInstanceTypeKey(instanceType, region, account),
-                  instanceTypeAttributes,
-                  Collections.emptyMap()));
-        }
-      }
+    return new DefaultCacheData(
+        "metadata",
+        objectMapper.convertValue(metadataAttributes, ATTRIBUTES),
+        Collections.emptyMap());
+  }
 
-      return new DefaultCacheResult(cacheResults, evictions);
-    } catch (Exception ex) {
-      throw new RuntimeException(ex);
+  private List<InstanceTypeInfo> getInstanceTypes(AmazonEC2 ec2) {
+    final List<InstanceTypeInfo> instanceTypeInfoList = new ArrayList<>();
+    final DescribeInstanceTypesRequest request = new DescribeInstanceTypesRequest();
+    while (true) {
+      final DescribeInstanceTypesResult result = ec2.describeInstanceTypes(request);
+      instanceTypeInfoList.addAll(result.getInstanceTypes());
+      if (result.getNextToken() != null) {
+        request.withNextToken(result.getNextToken());
+      } else {
+        break;
+      }
     }
-  }
 
-  Optional<String> getEtagHeader(HttpResponse response) {
-    return Optional.ofNullable(response)
-        .map(r -> r.getFirstHeader("ETag"))
-        .map(Header::getElements)
-        .filter(e -> e.length > 0)
-        .map(e -> e[0].getName());
+    return instanceTypeInfoList;
   }
 
   @Override
   public String getAgentType() {
-    return getClass().getSimpleName() + "/" + region;
+    return String.format("%s/%s/%s", getClass().getSimpleName(), account.getName(), region);
```

Comment:
With info cached with the previous agent type get evicted?  Or does this change make it so that old info will stick around?

Reply:
You suspected right, I see both the old and new keys in my redis cache. 
<img width="1729" alt="image" src="https://user-images.githubusercontent.com/3614196/186022197-b4bdba98-4fa1-4e71-b4a6-e18a6216a992.png">

Old key: `com.netflix.spinnaker.clouddriver.aws.provider.AwsInfrastructureProvider:AmazonInstanceTypeCachingAgent/eu-central-1:relationships`
New key: `com.netflix.spinnaker.clouddriver.aws.provider.AwsInfrastructureProvider:AmazonInstanceTypeCachingAgent/my-aws-devel-acct/eu-central-1:relationships`

I don't see a default TTL in Redis or Clouddriver config. As a result the old keys live forever until the redis database is flushed. This is probably not a common occurence, but would you know of a solution to cleanup the old keys? 

The only simple option I can think of is to keep the old agent type without the account name for the sake of backwards compatibility.

If we want to maintain the correctness of agent type and if a manual step before using these changes is okay, the release notes could include some commands to clean up the old keys?

Reply:
I think it's important to include the account name, since the info is now account specific.  I suppose we're ending up with a bunch of duplicate information since I can't imagine instance types vary much across accounts.  Not sure what to do about that though.  Probably nothing.

One reason I think including the account name is important is to support dynamic accounts.  When an account disappears, for code to clean up information associated with the account, I suspect the name needs to be there.

I think a release note to remove the old info is ok.

Reply:
Thanks for https://github.com/spinnaker/spinnaker.io/pull/230.